File: bench/cases/matmul/matmul.js
// matmul: naive NxN integer matrix multiply — a memory-bound triple-nested
// loop. Matrices are stored flat (row-major, i*N+j) in Int32Array so every
// language shares the same layout. checksum = sum of C mod MOD.
"use strict";

const N = 512;
const MOD = 1000003;
const expected = 434781;

const a = new Int32Array(N * N);
const b = new Int32Array(N * N);
const c = new Int32Array(N * N);
for (let i = 0; i < N; i++) {
  for (let j = 0; j < N; j++) {
    a[i * N + j] = (i + j) % 100;
    b[i * N + j] = (i * j) % 100;
  }
}

for (let i = 0; i < N; i++) {
  for (let j = 0; j < N; j++) {
    let s = 0;
    for (let k = 0; k < N; k++) {
      s += a[i * N + k] * b[k * N + j];
    }
    c[i * N + j] = s;
  }
}

let sum = 0;
for (let idx = 0; idx < N * N; idx++) {
  sum = (sum + c[idx]) % MOD;
}
const checksum = sum;
if (checksum !== expected) {
  process.stderr.write(`MISMATCH got=${checksum} want=${expected}\n`);
  process.exit(1);
}
process.exit(0);

File: bench/cases/matmul/matmul.go
package main

// matmul: naive NxN integer matrix multiply — a memory-bound triple-nested
// loop. Matrices are stored flat (row-major, i*n+j) so every language shares
// the same layout and access pattern. checksum = sum of all C[i][j] mod MOD.

import (
	"fmt"
	"os"
)

const n = 512
const mod int64 = 1000003
const expected int64 = 434781

func main() {
	a := make([]int32, n*n)
	b := make([]int32, n*n)
	c := make([]int32, n*n)
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			a[i*n+j] = int32((i + j) % 100)
			b[i*n+j] = int32((i * j) % 100)
		}
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			var s int32 = 0
			for k := 0; k < n; k++ {
				s += a[i*n+k] * b[k*n+j]
			}
			c[i*n+j] = s
		}
	}

	var sum int64 = 0
	for idx := 0; idx < n*n; idx++ {
		sum = (sum + int64(c[idx])) % mod
	}
	checksum := sum
	if expected < 0 {
		fmt.Fprintf(os.Stderr, "checksum=%d\n", checksum)
		os.Exit(0)
	}
	if checksum != expected {
		fmt.Fprintf(os.Stderr, "MISMATCH got=%d want=%d\n", checksum, expected)
		os.Exit(1)
	}
	os.Exit(0)
}
